File: packages/shared/src/cognitive/investor-profile.ts
/**
 * ART-PROFILE — InvestorProfile (RFC-008).
 * Declared ≠ Observed: never auto-merge into Policy.
 * Catálogo: tabla investor_profiles; cuentas referencian activeProfileId.
 */

import type { SuggestablePolicyTemplateId } from './suggest-policy-from-declared.js';

export type ProfileHorizon =
  | 'intraday'
  | 'swing'
  | 'position'
  | 'long_term';

export type RiskTolerance = 'low' | 'moderate' | 'high';

export type ExperienceLevel = 'novice' | 'intermediate' | 'advanced' | 'professional';

export type ProfileUpdatedBy = 'user' | 'system_observation' | 'hybrid';

/** Quién dice ser el inversor (cuestionario). */
export interface DeclaredInvestorProfile {
  horizon: ProfileHorizon;
  objectives: string[];
  riskTolerance: RiskTolerance;
  experience: ExperienceLevel;
  maxAcceptableLossPct?: number;
  notes?: string;
}

/**
 * Comportamiento medido — solo observación.
 * Nunca reescribe Declared ni TradingPolicy.
 */
export interface ObservedInvestorProfile {
  sampleTradeCount: number;
  impulsivityScore?: number;
  overtradingScore?: number;
  disciplineScore?: number;
  /** true si el comportamiento diverge del perfil declarado / policy activa */
  divergesFromDeclared: boolean;
  divergesFromPolicy: boolean;
  lastObservedAt?: string;
  notes?: string[];
}

/** Ficha de catálogo ART-PROFILE (tabla investor_profiles). */
export interface InvestorProfileV1 {
  artifactType?: 'ART-PROFILE';
  schemaVersion?: '1.0.0';
  profileId: string;
  version: string;
  name: string;
  userId?: string | null;
  declared: DeclaredInvestorProfile;
  observed?: ObservedInvestorProfile | null;
  suggestedPolicyTemplateId: SuggestablePolicyTemplateId | string;
  selectedPolicyTemplateId: SuggestablePolicyTemplateId | string;
  updatedBy: ProfileUpdatedBy | string;
  updatedAt: string;
  createdAt: string;
}

export interface ProfilePolicyDivergence {
  profileId: string;
  policyId: string;
  divergesFromDeclared: boolean;
  divergesFromPolicy: boolean;
  messages: string[];
}

File: packages/py/analytics/src/bolsa_analytics/cognitive/investor_profile.py
"""ART-PROFILE — Declared ≠ Observed (RFC-008 D1)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

ProfileHorizon = Literal["intraday", "swing", "position", "long_term"]
RiskTolerance = Literal["low", "moderate", "high"]
ExperienceLevel = Literal["novice", "intermediate", "advanced", "professional"]
ProfileUpdatedBy = Literal["user", "system_observation", "hybrid"]


@dataclass(frozen=True, slots=True)
class DeclaredInvestorProfile:
    horizon: ProfileHorizon
    objectives: tuple[str, ...]
    risk_tolerance: RiskTolerance
    experience: ExperienceLevel
    max_acceptable_loss_pct: float | None = None
    notes: str | None = None


@dataclass(frozen=True, slots=True)
class ObservedInvestorProfile:
    sample_trade_count: int
    diverges_from_declared: bool
    diverges_from_policy: bool
    impulsivity_score: float | None = None
    overtrading_score: float | None = None
    discipline_score: float | None = None
    last_observed_at: str | None = None
    notes: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class InvestorProfile:
    profile_id: str
    version: str
    name: str
    declared: DeclaredInvestorProfile
    updated_by: ProfileUpdatedBy
    updated_at: str
    created_at: str
    artifact_type: str = "ART-PROFILE"
    schema_version: str = "1.0.0"
    account_id: str | None = None
    observed: ObservedInvestorProfile | None = None
    suggested_policy_template_id: str | None = None

    def to_dict(self) -> dict[str, Any]:
        d = self.declared
        obs = self.observed
        return {
            "artifactType": self.artifact_type,
            "schemaVersion": self.schema_version,
            "profileId": self.profile_id,
            "version": self.version,
            "accountId": self.account_id,
            "name": self.name,
            "declared": {
                "horizon": d.horizon,
                "objectives": list(d.objectives),
                "riskTolerance": d.risk_tolerance,
                "experience": d.experience,
                "maxAcceptableLossPct": d.max_acceptable_loss_pct,
                "notes": d.notes,
            },
            "observed": None
            if obs is None
            else {
                "sampleTradeCount": obs.sample_trade_count,
                "impulsivityScore": obs.impulsivity_score,
                "overtradingScore": obs.overtrading_score,
                "disciplineScore": obs.discipline_score,
                "divergesFromDeclared": obs.diverges_from_declared,
                "divergesFromPolicy": obs.diverges_from_policy,
                "lastObservedAt": obs.last_observed_at,
                "notes": list(obs.notes),
            },
            "suggestedPolicyTemplateId": self.suggested_policy_template_id,
            "updatedBy": self.updated_by,
            "updatedAt": self.updated_at,
            "createdAt": self.created_at,
        }
